### ISSUE TITLE
feat: PyTorch-style GradientTape with op-level backward recording

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1057]
## Summary

Rewrite `GradientTape<T>` to support PyTorch-style autograd where tensor operations automatically record themselves to an active tape during forward execution, enabling exact gradient computation without a separate `ExportComputationGraph` path.

## Problem

The current autodiff system has two parallel forward paths:
- `Forward()` — does the actual computation
- `ExportComputationGraph()` — manually rebuilds the computation as a graph

These diverge. DiffusionResBlock's graph drops time conditioning. VAEEncoder/VAEDecoder graphs are stubs that throw `NotSupportedException`. The SPSA fallback exists because the graph path is unreliable. This is fundamentally unfixable without eliminating the dual-path architecture.

### GradientTape API

```csharp
// Thread-local active tape (like PyTorch's default grad mode)
public sealed class GradientTape<T> : IDisposable
{
    [ThreadStatic] private static GradientTape<T>? _current;
    public static GradientTape<T>? Current => _current;

    // Record an operation with its backward function
    public void RecordOp(
        string opName,
        Tensor<T>[] inputs,
        Tensor<T> output,
        Func<Tensor<T>, Tensor<T>[]> backward);

    // Compute gradients of target w.r.t. all watched tensors
    public Dictionary<Tensor<T>, Tensor<T>> Gradient(Tensor<T> loss);
}

// No-grad context for inference (avoids tape overhead)
public sealed class NoGradScope : IDisposable { ... }
```

### Engine ops record to tape

```csharp
public Tensor<T> TensorAdd<T>(Tensor<T> a, Tensor<T> b)
{
    var result = /* actual SIMD computation */;

    if (GradientTape<T>.Current is { } tape)
    {
        tape.RecordOp("add", [a, b], result,
            grad => [grad, grad]); // d(a+b)/da = 1, d(a+b)/db = 1
    }

    return result;
}
```

### Tensor parameter tracking

```csharp
// Mark tensors as requiring gradients (like PyTorch requires_grad=True)
public class Tensor<T>
{
    public bool RequiresGrad { get; set; }
    public Tensor<T>? Gradient { get; internal set; } // Accumulated gradient
}
```

### Step 1: Core GradientTape rewrite
- [ ] Implement `GradientTape<T>` with `RecordOp`, `Gradient`, thread-static `Current`
- [ ] Implement backward graph traversal with reverse topological sort
- [ ] Implement gradient accumulation for tensors used in multiple ops
- [ ] Implement `NoGradScope` for inference (disables tape recording)
- [ ] Add `RequiresGrad` and `Gradient` properties to `Tensor<T>`

### Step 2: Register backward functions for all Engine ops
Each op needs its backward function. Reference: PyTorch autograd function registry.

**Elementwise ops:**
- [ ] Add: `grad → (grad, grad)`
- [ ] Subtract: `grad → (grad, -grad)`
- [ ] Multiply: `grad → (grad * b, grad * a)`
- [ ] Divide: `grad → (grad / b, -grad * a / b²)`
- [ ] Negate: `grad → (-grad)`

**Scalar ops:**
- [ ] AddScalar: `grad → grad`
- [ ] MultiplyScalar: `grad → grad * scalar`
- [ ] DivideScalar: `grad → grad / scalar`

**Reduction ops:**
- [ ] Sum: `grad → broadcast(grad)`
- [ ] Mean: `grad → broadcast(grad / n)`

**Matrix ops:**
- [ ] MatMul(A, B): `grad → (grad @ B^T, A^T @ grad)`
- [ ] Transpose: `grad → transpose(grad)`

**Activation ops:**
- [ ] Tanh: `grad → grad * (1 - tanh(x)²)`
- [ ] Sigmoid: `grad → grad * σ(x) * (1 - σ(x))`
- [ ] ReLU: `grad → grad * (x > 0)`
- [ ] Swish/SiLU: `grad → grad * (σ(x) + x * σ(x) * (1 - σ(x)))`
- [ ] GELU: `grad → grad * GELU'(x)`
- [ ] Softmax: `grad → grad - sum(grad * softmax) * softmax` (Jacobian-vector product)

**Conv ops:**
- [ ] Conv2D: backward through im2col
- [ ] Conv2DTranspose: backward through col2im

**Normalization ops:**
- [ ] BatchNorm: backward through running stats
- [ ] GroupNorm: backward through group statistics
- [ ] LayerNorm: backward through layer statistics

### Step 3: Gradient checking test suite
- [ ] For each op, verify `tape.Gradient()` matches finite-difference approximation within tolerance
- [ ] Test gradient flow through multi-op chains (e.g., `tanh(matmul(x, w) + b)`)
- [ ] Test gradient accumulation when a tensor is used in multiple ops
- [ ] Test that `NoGradScope` prevents recording
- [ ] Test thread safety of `GradientTape.Current`

## Acceptance Criteria
- All registered ops pass gradient checking (autodiff vs finite-difference < 1e-5 relative error)
- No performance regression when tape is not active (inference path)
- Existing `ExportComputationGraph` still works (not yet removed — that's issue #3)

## References
- PyTorch autograd: https://pytorch.org/docs/stable/autograd.html
- JAX autodiff: https://jax.readthedocs.io/en/latest/autodiff-cookbook.html
- Baydin et al. "Automatic Differentiation in Machine Learning: a Survey" (2018)
